ui/src/app/pages/data/data-set/dataset-page.tsx

```diff
@@ -426,47 +444,100 @@ const DataSetPage = fp.flow(withUserProfile(), withCurrentWorkspace(), withUrlPa
 
     async componentDidMount() {
       const {namespace, id} = this.props.workspace;
-      const allPromises = [];
-      allPromises.push(this.loadResources());
-      if (this.editing) {
-        allPromises.push(dataSetApi().getDataSet(
-          namespace, id, this.props.urlParams.dataSetId).then((response) => {
-            this.setState({
-              dataSet: response,
-              includesAllParticipants: response.includesAllParticipants,
-              selectedConceptSetIds: response.conceptSets.map(cs => cs.id),
-              selectedCohortIds: response.cohorts.map(c => c.id),
-              selectedDomainValuePairs: response.domainValuePairs,
-              valuesLoading: true,
-            });
-            if (response.prePackagedConceptSet === PrePackagedConceptSetEnum.BOTH) {
-              this.setState({prePackagedSurvey: true, prePackagedDemographics: true});
-            } else if (response.prePackagedConceptSet === PrePackagedConceptSetEnum.DEMOGRAPHICS) {
-              this.setState({prePackagedDemographics: true});
-            } else if (response.prePackagedConceptSet === PrePackagedConceptSetEnum.SURVEY) {
-              this.setState({prePackagedSurvey: true});
+      const resourcesPromise = this.loadResources();
+      if (!this.editing) {
+        return;
+      }
+
+      const [, dataSet] = await Promise.all([
+        resourcesPromise,
+        dataSetApi().getDataSet(namespace, id, this.props.urlParams.dataSetId)
+      ]);
+
+      const selectedPrepackagedConceptSets = this.apiEnumToPrePackageConceptSets(dataSet.prePackagedConceptSet);
+      this.setState({
+        dataSet,
+        includesAllParticipants: dataSet.includesAllParticipants,
+        selectedConceptSetIds: dataSet.conceptSets.map(cs => cs.id),
+        selectedCohortIds: dataSet.cohorts.map(c => c.id),
+        selectedDomainValuePairs: dataSet.domainValuePairs,
+        selectedDomains: this.getDomainsFromConceptSets(dataSet.conceptSets, selectedPrepackagedConceptSets),
+        selectedPrepackagedConceptSets,
+      });
+    }
+
+    async componentDidUpdate({}, prevState: State) {
+      // After a state update, first check whether any new domains have been added.
+      const newDomains = Array.from(this.state.selectedDomains)
+          .filter(d => !prevState.selectedDomains.has(d));
+      if (!newDomains.length) {
+        return;
+      }
+
+      // TODO: There is a lot of complexity here around loading domain values
+      // which is static data for a given CDR version. Consider refactoring this
```

Comment:
Is there a ticket for the refactor here? If not, we probably should file one.

Reply:
Filed https://precisionmedicineinitiative.atlassian.net/browse/RW-4426